Automattic/woocommerce-payments
# Store the customer currencies data as an option

Fixes #7000

#### Changes proposed in this Pull Request

This PR aims to improve the `register_customer_currencies` performance by replacing the expensive calculation with an option that is updated on every new order. The expensive `get_all_customer_currencies` call will be called only the first time, and then its value will be used to set an option.

#### Benchmark

The benchmark was done using `microtime()` to calculate the start and end times of the function execution.

##### Benchmark with 205 orders

<table>
    <thead>
        <tr>
            <th>#</th>
            <th>Previous</th>
            <th>Current</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1</td>
            <td><code>0.001592874526977539μs</code></td>
            <td><code>0.000027179718017578125μs</code></td>
        </tr>
        <tr>
            <td>2</td>
            <td><code>0.0019979476928710938μs</code></td>
            <td><code>0.000025987625122070312μs</code></td>
        </tr>
       <tr>
            <td>3</td>
            <td><code>0.0030279159545898438μs</code></td>
            <td><code>0.000030040740966796875μs</code></td>
        </tr>
       <tr>
            <td>4</td>
            <td><code>0.0017621517181396484μs</code></td>
            <td><code>0.000024080276489257812μs</code></td>
        </tr>
       <tr>
            <td>5</td>
            <td><code>0.00202178955078125μs</code></td>
            <td><code>0.000032901763916015625μs</code></td>
        </tr>
        <tr>
            <td>Avg</td>
            <td><code>0.002080535888672μs</code></td>
            <td><code>0.000028038024902μs</code></td>
        </tr>
        <tr>
            <td>Diff</td>
            <td colspan="2">~74x speedup</td>
        </tr>
    </tbody>
</table>

##### Benchmark with 5723 orders

<table>
    <thead>
        <tr>
            <th>#</th>
            <th>Previous</th>
            <th>Current</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1</td>
            <td><code>0.5904631614685059μs</code></td>
            <td><code>0.00003600120544433594μs</code></td>
        </tr>
        <tr>
            <td>2</td>
            <td><code>0.6487610340118408μs</code></td>
            <td><code>0.00003600120544433594μs</code></td>
        </tr>
       <tr>
            <td>3</td>
            <td><code>0.8029019832611084μs</code></td>
            <td><code>0.000032901763916015625μs</code></td>
        </tr>
       <tr>
            <td>4</td>
            <td><code>0.6125071048736572μs</code></td>
            <td><code>0.00003409385681152344μs</code></td>
        </tr>
       <tr>
            <td>5</td>
            <td><code>0.5736191272735596μs</code></td>
            <td><code>0.000031948089599609375μs</code></td>
        </tr>
        <tr>
            <td>Avg</td>
            <td><code>0.642</code></td>
            <td><code>0.00003416</code></td>
        </tr>
        <tr>
            <td>Diff</td>
            <td colspan="2">~18,793x speedup</td>
        </tr>
    </tbody>
</table>









#### Testing instructions





* Checkout to this branch
* Go to the analytics page `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
  * The new option will be set on this page load
* After the page loads, open a new tab on PhpMyAdmin `http://localhost:8083/` and run the following query:
```sql
SELECT * FROM `wp_options` WHERE `option_name` LIKE "wcpay_multi_currency_stored_customer_currencies";
```
* Go to the Multi-currency settings page `/wp-admin/admin.php?page=wc-settings&tab=wcpay_multi_currency` and enable a new currency
* Go to your shop and perform the checkout using the currency you've just added
* Go to the PhpMyAdmin tab and refresh the query
* Check the array on `option_value`, it should contain the currency code
* Go to the analytics page `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
* Check the "Customer currency" select – it should also contain the currency code option

-------------------


**Post merge**
